[4.0] A zero cost huge performance win

dgrammatiko opened the pull request:
Pull Request for Issue # .

### Summary of Changes
- Add a small script to the arsenal of NPM tools so that scripts and stylesheets can be gzipped

### WHY?
Performance is really crucial these days, especially since Google is accounting it as a metric to rank the pages. Joomla is not meeting the today needs. Prove that in the HTTP archive the Joomla websites rank second to last (WIX): 

![](https://discourse-cdn-sjc2.com/standard17/uploads/httparchive/original/2X/7/79f6dc42c93f657e9112174daa6c19459b3725bf.png)

Link: https://discuss.httparchive.org/t/cms-performance/1468

The main reason for this is that Joomla has ALL the scripts in the head of the document and also a huge CSS file that blocks the rendering. But these cannot seem to be ok to be fixed for various reasons. So let's try to get some small wins from various tweaks.
In this PR we provide a way to gzip the script and stylesheet files ONCE (probably in the build of the installable zip files). This will free up some processor cycles in every request since the files are already gzipped so Apache or Nginx don't have to stupidly repeat the process for every request. Simple, efficient, performant. This has been my way of handling the assets for many years with 0 problems, so it's also battle tested somehow.


### How to
Run the command 
```bash
node build.js --build-gzip
```
in the root of your Joomla installation.

If you're in Apache follow the instructions here: https://httpd.apache.org/docs/2.4/mod/mod_deflate.html#page-header
Basically: 
```
<IfModule mod_headers.c>
    # Serve gzip compressed CSS files if they exist 
    # and the client accepts gzip.
    RewriteCond "%{HTTP:Accept-encoding}" "gzip"
    RewriteCond "%{REQUEST_FILENAME}\.gz" -s
    RewriteRule "^(.*)\.css" "$1\.css\.gz" [QSA]

    # Serve gzip compressed JS files if they exist 
    # and the client accepts gzip.
    RewriteCond "%{HTTP:Accept-encoding}" "gzip"
    RewriteCond "%{REQUEST_FILENAME}\.gz" -s
    RewriteRule "^(.*)\.js" "$1\.js\.gz" [QSA]


    # Serve correct content types, and prevent mod_deflate double gzip.
    RewriteRule "\.css\.gz$" "-" [T=text/css,E=no-gzip:1]
    RewriteRule "\.js\.gz$" "-" [T=text/javascript,E=no-gzip:1]


    <FilesMatch "(\.js\.gz|\.css\.gz)$">
      # Serve correct encoding type.
      Header append Content-Encoding gzip

      # Force proxies to cache gzipped & 
      # non-gzipped css/js files separately.
      Header append Vary Accept-Encoding
    </FilesMatch>
</IfModule>
```

### Documentation Changes Required
To production team: If you accept this you need to run the command above before each realise (can be automated, I guess) and also you need to provide a solid guide based on the info above so Admins can tweak the apache/nginx configurations. Also one thing that might needs to be considered is that the size of the installable/updates will increase by a good amount of Mbs.
Lastly if this will ever be consider I also propose that you force the extensions in the JED to also provide gzipped files for all their scripts/stylesheets.

And a final word: even if you reject to actually formally follow the proposed path, you can at least provide the tool so people that want to provide decent experiences at least have the means.

dgrammatiko committed:
Add gzip to the arsenal of tools

dgrammatiko committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into 4.0-dev-gzip-once

* '4.0-dev' of github.com:joomla/joomla-cms: (40 commits)
  [4.0] ESLint javascript scripts or fail (#23363)
  Fix javascript lint (#23381)
  Fix codestyle
  Fix codestyle
  [com_tags] Image caption in Tags view (#20648)
  Singular not plural (#20751)
  [com_menus] - fix php warning (#18471)
  Fix stale session data wrongly overriding configuration file when editing global configuration (#20590)
  Update strings (#20591)
  Display 404 page if banner not found (#20664)
  JTableAsset::loadByName(): replace two sql queries by one (#20671)
  Replace UCMType::getType() by UCMType::getTypeByAlias($this->alias) (#20672)
  Verb wrong tense (#20708)
  [Newsfeeds] Use item link instead of guid. (#20717)
  [CS] fix some doc comment types (round 2) (#20647)
  Joomla\CMS\Categories\Categories fix (#20680)
  Don't try to json_decode() a null value (#20675)
  [fix] Error decoding JSON data: Syntax error (#20663)
  Bump Composer dependencies (#20583)
  CodeMirror updated to 5.38.0 (#20636)
  ...

dgrammatiko committed:
only the minified files

richard67 commented:
@dgrammatiko I have done this method too before my hoster finally enabled mod_deflate for shared hosting, but I did not use extension `gz`, I used `jgz` for reasons stated here:

- [https://stackoverflow.com/questions/1235116/safari-and-gzip](https://stackoverflow.com/questions/1235116/safari-and-gzip)
- [http://blog.bigsmoke.us/2012/01/16/safari-ignores-content-type-for-gz-suffix](http://blog.bigsmoke.us/2012/01/16/safari-ignores-content-type-for-gz-suffix)

I suggest that you check the information linked above, and if still relevant consider to use `jgz` instead of `gz`.

dgrammatiko commented:
I don't think any of these apply here:
- safari v4 is outdated
- I'm not proposing to call/use the file.js.gz directly, this is done automagically by either Apache or Nginx.

Some explanation how all this works:
- User makes a request to our site
- Joomla creates the html and sends it back
- Browser parses the page, finds all the assets and starts making requests
- Apache receives the request
- Checks if the header that is send by the browser got the gzip deflate : eg `Accept-Encoding: br, gzip, deflate`
- If gzip is supported **apache** through the snippet provided above will switch off the automatic gzipping of the file and serve the already stored one.

The process power released by not having to gzip files in every request is enough to raise the maximum number of concurrent client. Not to mention that this is **environmentally friendly programmingⓇ**

BTW the provided default snippet by apache will **always** switch the extensions, I guess this needs an if statement to check if the file actually exists...

richard67 commented:
> I'm not proposing to call/use the file.js.gz directly, this is done automagically by either Apache or Nginx.
I know.

> Some explanation how all this works:
I know.

> BTW the provided default snippet by apache will always switch the extensions, I guess this needs an if statement to check if the file actually exists...

Yes, there should be a `RewriteCond %{REQUEST_FILENAME}.gz -f` at the end of the rewrite conditions chain for each of the file types.

richard67 commented:
Oh, I just see there is already a `RewriteCond "%{REQUEST_FILENAME}\.gz" -s`. That should do already the job.

dgrammatiko commented:
Here is my nginx conf:
```
# gzip
gzip_static  on;
gzip on;
gzip_vary on;
gzip_proxied any;
gzip_comp_level 6;
gzip_types text/plain text/css text/xml application/json application/javascript application/xml+rss application/atom+xml image/svg+xml application/font-woff  application/font-woff2;
```

The only extra part needed to work is `gzip_static  on;`
Ref: http://nginx.org/en/docs/http/ngx_http_gzip_static_module.html

ggppdk commented:
very nice to see this pr,

i understand how this works,
i have used it myself in some cases

but i want to ask regarding the case of having apache as web server
why this PR does not include the suggested changes inside htaccess.txt too ?
i mean htaccess.txt is used in new installations so why not include the suggested modification ?
are we expecting issues with contents of 
```
<IfModule mod_headers.c>
    # Serve gzip compressed CSS files if they exist 
    # and the client accepts gzip.
...
</IfModule>
```

dgrammatiko commented:
@wilsonge can we get a decision from the production team on this one?

PS. The drone is failing, probably because the linux is 32bit and the zopfli script exists only for 64bit systems.

By the way, for anyone care to dig into the used technology here that is zopfli. Zopfli is compression scripts from Google, basically an improved gzip: https://github.com/google/zopfli

dgrammatiko committed:
Merge branch '4.0-dev' into 4.0-dev-gzip-once

dgrammatiko committed:
Refactor it

dgrammatiko committed:
oops

dgrammatiko committed:
Fix the doc blocks remove any dead code

dgrammatiko committed:
Partially restore watch

Update all dev dependencies

dgrammatiko committed:
CS

+ keep the language order in the error pages constant (eg no falsy changes marked)

dgrammatiko committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into 4.0-dev-tools-refactoring

* '4.0-dev' of github.com:joomla/joomla-cms:
  [4.0] Housekeeping (#23600)
  [4.0] com_finder: Filtering common & numeric words and language fixes and params (#21037)

dgrammatiko committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into 4.0-dev-tools-refactoring

* '4.0-dev' of github.com:joomla/joomla-cms:
  [4.0] Correcting icons display in debug bar (#23732)

dgrammatiko committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into 4.0-dev-tools-refactoring

* '4.0-dev' of github.com:joomla/joomla-cms:
  [4.0] Fix database name validation for MySQL in installation app (#23704)
  [4.0] Normalize sample data plugins to use Joomla default workflow (#23719)
  [4.0] Correcting success colour after installing sample data (#23726)

dgrammatiko committed:
Merge branch '4.0-dev' into 4.0-dev-tools-refactoring

dgrammatiko committed:
Merge branch '4.0-dev-tools-refactoring' of github.com:dgrammatiko/joomla-cms into 4.0-dev-gzip-once

* '4.0-dev-tools-refactoring' of github.com:dgrammatiko/joomla-cms:
  Clean the node modules in the media folder on install (#22860)
  Remove libraries/cms/html/language/en-GB/en-GB.jhtmldate.ini (#23653)
  [4.0] Fix save contact (#23688)
  [4.0] Fix database name validation for MySQL in installation app (#23704)
  [4.0] Normalize sample data plugins to use Joomla default workflow (#23719)
  [4.0] Correcting success colour after installing sample data (#23726)
  [4.0] Correcting icons display in debug bar (#23732)
  [4.0] Housekeeping (#23600)
  CS
  Partially restore watch
  Fix the doc blocks remove any dead code
  oops
  [4.0] com_finder: Filtering common & numeric words and language fixes and params (#21037)
  Refactor it
  Set inital workflow for the uncategories category to "use default" (#23701)
  com_finder categories plugin needs to check right extension element (#20658)
  [4.0][a11y]Admin module logged button (#23277)
  Remove myself from most CODEOWNERS paths (#23707)

# Conflicts:
#	build.js
#	package.json

dgrammatiko committed:
Joomla says hello to Web assembly

dgrammatiko commented:
Redid this using WASM so no further changes required.
Attention this is now based on the new code for the build tools...

HLeithner commented:
is there a reason why we don't create brotli compressed files? gzip is done by the moste webservers/loadbalancers/caches on the fly but brotli support is still very limited.

dgrammatiko commented:
@HLeithner we could do that as well. The idea here is that these are repeated tasks from the web server which require processor power, so by providing them ready for delivery it's only a case of reading the file and streaming it...

FWIW I like the idea that Joomla will also provide prepared Brotli files

dgrammatiko committed:
Also Brotli

dgrammatiko commented:
@HLeithner it seems zopfli is able to achieve the same level of compression as brotli: 
<img width="1546" alt="screenshot 2019-02-03 at 22 42 00" src="https://user-images.githubusercontent.com/3889375/52183144-36bfd580-2805-11e9-8fad-268c22465daf.png">

dgrammatiko committed:
oops

dgrammatiko committed:
Wasm Brotlli is flaky

dgrammatiko committed:
Update + htaccess changes + brotli configurable

dgrammatiko committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into 4.0-dev-gzip-once

* '4.0-dev' of github.com:joomla/joomla-cms:
  Some webcomponents already pass linting (#23789)
  Fix running npm install and npm ci (#23788)
  [4.0][NO CACHE] Refactor the build tools (#23723)
  Fix saving content menu items (#23716)
  [4.0] remove duplicate string (#23771)
  Remove the slider button (#23388)
  [4.0] Set default workflow for protected workflows (#23607)
  modified_time check (#23689)
  [4.0] FIX SQL int default value to  numeric (#23669)
  Clean up hard coded usage of state values (#23330)
  [4.0] - fix save article (#23638)
  [4.0] JS error when closing a item and going back to a list view (#23642)
  Fixes plugins ordering (#23731)
  [4.0] [a11y] Make ActionButton Featured accessible (#23718)
  Remove extra type attribute (#23753)
  [4.0] Remove unused string (#23750)
  Fix stats alert information (#23727)
  [4.0] Remove deprecated functions isSite and isAdmin (#22492)

# Conflicts:
#	.eslintignore
#	build.js
#	package-lock.json
#	package.json

dgrammatiko committed:
fixes

dgrammatiko committed:
Update .gitignore

wilsonge committed:
Merge branch '4.0-dev' into 4.0-dev-gzip-once

wilsonge committed:
Merge branch '4.0-dev' into 4.0-dev-gzip-once

wilsonge commented:
@dgrammatiko you've accidentally committed the templates compiled css (and minified css) files here

dgrammatiko committed:
upd

dgrammatiko committed:
remove the compiled files

dgrammatiko committed:
again, maybe

dgrammatiko committed:
correction to package-lock.json

dgrammatiko commented:
@wilsonge can you try to build (locally) a 4.0-dev release so you can verify that this is actually working?

wilsonge commented on build/build.php:
```diff
@@ -204,8 +204,12 @@ function clean_composer(string $dir)
 system('npm install --unsafe-perm', $npmReturnCode);
 
+// Create gzipped version of the static assets
+system('npm run gzip');
```
This should go after the npm install return code check

dgrammatiko replied:
done

wilsonge commented on build/build.php:
```diff
@@ -204,8 +204,12 @@ function clean_composer(string $dir)
 	exit(1);
 }
 
+system('npm install @gfx/zopfli iltorb --unsafe-perm', $npmReturnCode);
```
these should be dev dependencies. All our npm stuff needs to end up in the package-lock

dgrammatiko replied:
So my idea here was that this is a script that will only be executed by the dev that will create a release thus it makes 0 sense to bloat the rest of the package.json with 2 more dependencies...
I'm trying to keep npm as fast as possible so ppl with low end devices don't have to wait too long whenever they run npm install

Anyways if you think that this is not a good idea I can reverse it

mbabker replied:
It's better if running from git is as close to running from a production package as possible.  The more differences there are between a git install and a ZIP install, the more potentials there are for problems to exist in release packages that won't exist in the repo.  And there are already enough of those headaches in 3.x with everything being in the repo (ref the releases where a file was added at a spot that was previously deleted and the deleted file wasn't removed from the deleted files array as a prime example).

dgrammatiko replied:
Ok, let's go back to the dependencies

dgrammatiko committed:
Added zopfli to the dependencies

dgrammatiko committed:
order

dgrammatiko committed:
Remove Brotli command, etc for the time

dgrammatiko committed:
comments

wilsonge commented:
```
Processing: /Users/george/Sites/joomla-cms/media/vendor/webcomponentsjs/js/webcomponents-sd.min.js
```

After this file it hung. I had to command + c to close the task

dgrammatiko commented:
@wilsonge it didn't hang it just takes very long to make all those .gz files. Try adding a console.log to see the files processing

dgrammatiko commented:
And for comparison here are the minified, normal gzip from a server, and the zopfli sizes:

56.658 bytes, 21.462 bytes, 17.881 bytes

<img width="820" alt="screenshot 2019-02-28 at 22 42 50" src="https://user-images.githubusercontent.com/3889375/53600687-71215600-3baa-11e9-9fd8-bbe4f79134f4.png">

dgrammatiko committed:
CS

wilsonge committed:
Merge branch '4.0-dev' into 4.0-dev-gzip-once

wilsonge commented:
Seems good to me. Thanks!
